.github/workflows/CI-tag-orchestrator.yml

```
@@ -5,54 +5,132 @@ on:
     tags:
       - "[0-9]+.[0-9]+.[0-9]+-[0-9]+.[0-9]+.[0-9]+*"
   workflow_dispatch:
+    inputs:
+      git_tag:
+        type: string
+    secrets:
+      DOCKER_HUB_USERNAME:
+        required: true
+      DOCKER_HUB_TOKEN:
+        required: true
 
 env:
   RUST_BACKTRACE: 1
 
 jobs:
-  execute-system-tests:
-    name: Execute System Tests
-    uses: HorizenLabs/zkVerify-qa/.github/workflows/CI-e2e-tests.yml@main
+  env-setup:
+    uses: ./.github/workflows/CI-build.yml
     with:
-      zkverify_version: ${{ github.ref_name }}
-    secrets:
-      DOCKER_HUB_USERNAME: ${{ secrets.DOCKER_HUB_USERNAME }}
-      DOCKER_HUB_TOKEN: ${{ secrets.DOCKER_HUB_TOKEN }}
-      QA_SLACK_WEBHOOK_URL: ${{ secrets.QA_SLACK_WEBHOOK_URL }}
-      GH_TOKEN: ${{ secrets.GH_TOKEN }}
-
-  build-and-publish:
-    name: Build and Publish Docker Image
+      CACHING_MODE: "CACHE_RESTORE"
+  
+  build:
     runs-on: ubuntu-latest
-    needs: execute-system-tests # Wait for e2e tests to finish
+    needs: env-setup
+
+    outputs:
+      artifact_name_solo: ${{ steps.create-tar-file.outputs.tar_file_name_solo }}
+      artifact_name_relay: ${{ steps.create-tar-file.outputs.tar_file_name_relay }}
+
     steps:
-      - name: Git checkout
-        uses: actions/checkout@v4
+      - uses: actions/checkout@v4
         with:
-          ref: ${{ github.ref }}
+          ref: ${{ github.event.inputs.git_tag || github.ref }}
           fetch-depth: 0
 
-      - name: Download zkVerify Docker image artifact
-        uses: actions/download-artifact@v4
+      - name: Set up deps cache
+        run: mkdir deps
+
+      - name: Restore common cache
+        if: ${{ !env.ACT }}
+        uses: actions/cache/restore@v3
+        with:
+          path: |
+            ~/.cargo/bin/
+            ~/.cargo/registry/index/
+            ~/.cargo/registry/cache/
+            ~/.cargo/git/db/
+            deps/
+          key: common-cache
+      
+      - name: Restore build cache
+        if: ${{ !env.ACT }}
+        uses: actions/cache/restore@v3
         with:
-          name: zkverify-image
-          path: ./
+          path: |
+            target/
+          key: build-cache
       
-      - name: Build/Publish Docker image
-        env:
-          RELEASE_BRANCH: release
-          MAINTAINERS_KEYS: ${{ vars.MAINTAINERS_KEYS }}
-          DOCKER_HUB_USERNAME: ${{ secrets.DOCKER_HUB_USERNAME }}
-          DOCKER_HUB_TOKEN: ${{ secrets.DOCKER_HUB_TOKEN }}
+      - name: Verify release build
```

```suggestion
      - name: Verify restored build cache artifacts
```

How about this instead if I understand this correctly